tv_apps/smart-home/js/filter_manager.js

```diff
@@ -0,0 +1,104 @@
+/* global CardFilter, CardManager */
+
+(function(exports) {
+  'use strict';
+
+  var FilterManager = function() {};
+
+  FilterManager.prototype = {
+    init: function(options) {
+      this._cardListElem = options.cardListElem;
+      this._cardScrollable = options.cardScrollable;
+      this._createCardNode = options.createCardNode;
+      this._cardManager = options.cardManager;
+
+      this._cardFilter = new CardFilter();
+      this._cardFilter.start(document.getElementById('filter-tab-group'));
+      this._cardFilter.filter = CardManager.FILTERS.ALL.iconName;
+      this._filterChangedHandler = this.onFilterChanged.bind(this);
+      this._cardFilter.on('filterchanged', this._filterChangedHandler);
+
+      this._smartBubblesElem = document.getElementById('bubbles');
+
+      this._animationEndHandler = this.onAnimationEnd.bind(this);
+      this._smartBubblesElem.addEventListener('all-items-bubbled',
+        this._animationEndHandler);
+    },
+
+    uninit: function() {
+      this._cardFilter.off('filterchanged', this._filterChangedHandler);
+      this._smartBubblesElem.removeEventListener('all-items-bubbled',
+        this._animationEndHandler);
+    },
+
+    _filteredCardList: undefined,
+
+    _isFirstFrame: false,
+
+    _isFilterChanging: false,
+
+    _isBubbleSinking: function fm_isBubbleSinking() {
+      return (
+        this._cardListElem.getAttribute('smart-bubbles-direction') === 'down');
+    },
+
+    _performBubbleUp: function fm_performBubbleUp() {
+      if (this._isFirstFrame) {
+        this._isFirstFrame = false;
+        this._cardListElem.style.transition = 'none';
+        this._cardScrollable.scrollToFirstItem();
+        window.requestAnimationFrame(this._performBubbleUp.bind(this));
+      } else {
+        this._cardListElem.style.opacity = 1;
+        this._cardListElem.style.transition = undefined;
+        this._cardListElem.removeAttribute('smart-bubbles-direction');
+        this._smartBubblesElem.play(
+          document.querySelectorAll('#card-list > .card > .app-button'));
+      }
+    },
+
+    onAnimationEnd: function fm_onAnimationEnd() {
+      var that = this;
+      var filter = CardManager.getFilterByIconName(this._cardFilter.filter);
```

Comment:
If `IconName` is an attribute for `CardFilter`, which is a view controller, it would be a little bit strange to let `CardManger` to do its work based on a specific view parameter or icon. Can we consider moving getFilterByIconName to `CardFilter` or `FilterManager`?